src/PowerManager.vala

```diff
@@ -35,26 +37,39 @@ public class Power.PowerManager : Object {
         }
     }
 
-    public bool has_battery () {
-        if (upower.on_battery) {
-            return true;
-        };
-
-        var has_battery = false;
-        devices.foreach ((path, device) => {
-            if (device.device_type == BATTERY) {
-                has_battery = true;
-            }
-        });
-
-        return has_battery;
-    }
-
     public bool on_battery () {
         return upower.on_battery;
     }
 
     public bool has_lid () {
         return upower.lid_is_present;
     }
+
+    private void on_device_added (ObjectPath device_path) {
+        var device = new Device (device_path);
+
+        uint position = -1;
+        var found = devices.find_with_equal_func (device, (EqualFunc<Device>) Device.equal_func, out position);
+
+        if (!found) {
+            devices.append (device);
+
+            /*
+            * Need to verify power-supply before considering it a laptop battery.
+            * Otherwise it will likely be the battery for a device of an unknown type.
+            */
+            if (device.device_type == BATTERY && device.power_supply) {
+                has_battery = true;
+            }
+        }
+    }
+
+    private void on_device_removed (ObjectPath device_path) {
+        uint position = -1;
+        devices.find_with_equal_func (new Device (device_path), (EqualFunc<Device>) Device.equal_func, out position);
+
+        if (position != -1) {
+            devices.remove (position);
+        }
+    }
```

Comment:
Shouldn't `has_battery` be updated here as well?  Maybe maintain `n_battery` count and have `has_battery` return `n_battery > 0`?

Reply:
Looking at this again I'll save myself a lot of pain by having a separate list store for batteries and connected devices, so I simplified this a bit and now in mainview we just check if the list of batteries is > 0